packages/medusa/src/services/event-bus.ts

```diff
@@ -129,8 +149,16 @@ export default class EventBusService {
       throw new Error("Subscriber must be a function")
     }
 
-    const observers = this.observers_.get(event) ?? []
-    this.observers_.set(event, [...observers, subscriber])
+    const uniqueSubscriberId = `${event.toString()}-${ulid()}` // <- maybe we can do better?
```

Comment:
thought: would it make sense to add a third optional arg to this function that allows the caller to specify the subscriber id?

Reply:
The reasoning is that this random id will create issues when multiple instances could pick up the job. Each worker will assign a different id to the same subscriber, which means that if the instance that processes the retry is not the same as the instance that did the first (or any previous attempt), the "filter" mechanism will not be reliable.

If, however, the caller of `subscribe` can specify a unique global id, that would mitigate this issue.

Reply:
A good implementation will check if the provided id has already been used and either throw or replace the existing subscriber with the same id.

Reply:
I don't think the actual approach would cause any issues even with multiple instances. The job data is getting updated with the subscriber ids that have been completed. So each instance when filtering, are gathering the subscriber to run in there instance as long as the id is not part of the list on the job data. wdyt?

Reply:
The id set upon subscribing will probably kick in when we will move to a streaming approach. It will avoid handling the same event with the same subscriber in different instances.

Reply:
![IMG_1532](https://user-images.githubusercontent.com/25098370/210992770-0f8e7ac0-0ad4-4d50-9d50-16d61d2d4f6f.png)

Here is the actual behavior for the same event with 2 instances. Could you confirm @olivermrbl

Reply:
![IMG_1533](https://user-images.githubusercontent.com/25098370/210994434-3e8219a7-6458-45c6-9bf6-4738f0b49e2a.png)

Here is the actual behavior for the same event with 2 instances. Could you confirm @olivermrbl
s1...s6 correspond to uniq ids :D

Reply:
in regard to the staged job, the enqueur generates the events from them. therefore the above mechanism should kick in no? @srindom correct me If I missed something

Reply:
Yes, each subscriber will have unique ids. 

In the case above, the issue that @srindom addresses, and please correct me if I am wrong, is that S3 in instance 1 will be retried even though S6 in instance 2 is completed. And since S3 and S6 is the same subscriber, the job is processed twice. 

Therefore, it might make sense to be able to provide a global subscriber id upon subscribing to events.

Reply:
Ok i see now thanks 😅 indeed, and in that case would it tackle the fact that by default they are always handled multiple times at the moment. Which was not the scope of that feature. Also, the global id needs to be a static value and not randomly generated, my point is more in case of an instance failure.

Reply:
Wrt. a multi-instance scenario. Imagine you have:
- two instances
- two subscribers - Subscriber A and Subscriber B
- an event that gets retried - in this case, twice before succeeding

Since each instance randomly generates an id the `completed_ids` gets populated with values that cannot be cross-referenced across workers. I.e., Subscriber A has a different ID in Instance 1 and Instance 2 (for the example's sake `A:1` and `A:2`).

| Attempt | Instance 1                                                                                                 | Instance 2                                                                                                                                                                                                                                |
|---------|------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 1       | completed_ids: undefined<br /> ✅ Subscriber A:1<br /> ✅ Subscriber B:1<br /> 🛑 Subscriber C:1 - fails                        |                                                                                                                                                                                                                                           |
| 2       |                                                                                                            | completed_ids: [A:1, B:1]<br /> ✅ Subscriber A:2 - this is the same code as A:1 but since the id is generated at random the completed_ids can't be used for filtering<br /> ✅ Subscriber B:2 - this is the same code as B:1 <br /> 🛑 Subscriber C:2 - fails |
| 3       | completed_ids: [A:1, B:1, A:2, B:2] <br />⏭ Subscriber A:1 - skipped <br />⏭ Subscriber B:1 - skipped <br />✅ Subscriber C:1 |                                                                                                                                                                                                                                           |

Since we can't guarantee that it is the same instance that retries the event, the completed_ids may end up not being helpful for filtering.

Reply:
Yeah this behavior is what we already have at the moment. It is something that we wanted to tackle in another task related to duplicated handling. What I mean is that this PR was not scoped to fix that scenario as with or without this pr this is already what is happening. But oil told me that you already discussed that, and I am more than happy to see it introduced as part of that scope in order to get a first fix iteration on that subject :D